unit_tests/test_wait.py

```diff
@@ -79,6 +81,109 @@ def callback(arg1, arg2):
+        calls = []
+
+        def callback(arg1, arg2):
+            calls.append((arg1, arg2))
+            if len(calls) == 5:
+                return "what ever"
+            return False
+
+        def set_stop_in_timeout(ev: threading.Event):
+            time.sleep(1)
+            ev.set()
+
+        ev = threading.Event()
+        th = threading.Thread(target=set_stop_in_timeout, kwargs={"ev": ev})
+        th.start()
+        res = wait_for(callback, timeout=3, step=.5, arg1=1, arg2=3, throw_exc=False, stop_event=ev)
+        self.assertFalse(res)
+        self.assertTrue(len(calls) < 5)
+        th.join()
+
+    def test_04_stop_by_event_2(self):
+        calls = []
+
+        def callback(arg1, arg2):
+            calls.append((arg1, arg2))
+            if len(calls) == 5:
+                return "what ever"
+            return False
+
+        ev = threading.Event()
+        th = ThreadPoolExecutor(max_workers=1).submit(wait_for, func=callback, timeout=3,
+                                                      step=.5, arg1=1, arg2=3, throw_exc=False, stop_event=ev)
+        time.sleep(1)
+        ev.set()
+        res = th.result()
+        assert not th.exception()
+        self.assertFalse(res)
+        self.assertTrue(len(calls) < 5)
+
+    def test_04_stop_by_event_without_exception(self):
+        calls = []
+
+        def callback(arg1, arg2):
+            calls.append((arg1, arg2))
+            if len(calls) == 5:
+                return "what ever"
+            return False
+
+        def set_stop_in_timeout(ev: threading.Event):
+            time.sleep(1)
+            ev.set()
+
+        ev = threading.Event()
+        th = threading.Thread(target=set_stop_in_timeout, kwargs={"ev": ev})
+        th.start()
+        res = wait_for(callback, timeout=3, step=.5, arg1=1, arg2=3, throw_exc=True, stop_event=ev)
+        self.assertFalse(res)
+        self.assertEqual(len(calls), 3)
+        th.join()
+
+    def test_04_return_result_before_stop_event_and_wait_timeout(self):
+        calls = []
+
+        def callback(arg1, arg2):
+            calls.append((arg1, arg2))
+            if len(calls) == 3:
+                return "what ever"
+            return False
+
+        def set_stop_in_timeout(ev: threading.Event):
+            time.sleep(2)
+            ev.set()
+
+        ev = threading.Event()
+        th = threading.Thread(target=set_stop_in_timeout, kwargs={"ev": ev})
+        th.start()
+        res = wait_for(callback, timeout=3, step=.5, arg1=1, arg2=3, throw_exc=False, stop_event=ev)
+        self.assertEqual(res, "what ever")
+        self.assertEqual(len(calls), 3)
+        th.join()
+
+    def test_04_raise_by_timeout_before_set_event(self):
+        calls = []
+
+        def callback(arg1, arg2):
+            calls.append((arg1, arg2))
+            if len(calls) == 8:
+                return "what ever"
+            return False
+
+        def set_stop_in_timeout(ev: threading.Event):
+            time.sleep(4)
```

Comment:
This is unit test. It should not depend on the real time.
In good case it must be solved with the `mock`s.

Reply:
removed the time dependency